src/gui/layertree/qgslayertreeview.cpp

```diff
@@ -154,7 +159,16 @@ void QgsLayerTreeView::modelRowsInserted( const QModelIndex &index, int start, i
         if ( QgsLayerTreeEmbeddedWidgetProvider *provider = QgsGui::layerTreeEmbeddedWidgetRegistry()->provider( providerId ) )
         {
           QModelIndex index = layerTreeModel()->legendNode2index( legendNodes[i] );
-          setIndexWidget( index, provider->createWidget( layer, i ) );
+          QWidget *wdgt = provider->createWidget( layer, i );
+          // Since column is resized to contents, limit the expanded width of embedded widgets,
+          // if they are not already limited, e.g. have the defaut MAX value.
+          // Else, embedded widget may grow very wide due to large legend graphics.
+          // TODO: Max width could be a configured setting.
+          if ( wdgt->maximumWidth() == QWIDGETSIZE_MAX )
+          {
+            wdgt->setMaximumWidth( 250 );
```

Comment:
This won't scale nicely on hidpi -- can you use a font metric based max instead? (and remember to scale by Qgis::UI_SCALE_FACTOR

Reply:
I looked into that, but, for example, the Opacity slider is only part textual. How would I scale relative to hidpi for the associated widget as well?

Also, I looked into having the widget relative to the treeview viewport, but that was not pleasant at all.

Reply:
Could I just use the same text scale factor for the max width of the widget?

Reply:
What I usually do is work out how many 'X' characters fit in the desired 250 px size on low dpi, then use 

   QFontMetrics( font() ).width( 'X') * 20 * Qgis::UI_SCALING_FACTOR

(If the desired size is about 20 characters wide, etc)